Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/AppInstalls.hs

```diff
@@ -0,0 +1,81 @@
+{-
+ Copyright 2022-23, Juspay India Pvt Ltd
+
+ This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License
+
+ as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version. This program
+
+ is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+
+ or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more details. You should have received a copy of
+
+ the GNU Affero General Public License along with this program. If not, see <https://www.gnu.org/licenses/>.
+-}
+
+module Domain.Action.UI.AppInstalls
+  ( AppInstallsReq (..),
+    createAppInstallsDetails,
+  )
+where
+
+import qualified Domain.Types.AppInstalls as DAppInstalls
+import qualified Domain.Types.Merchant as DMerchant
+import Kernel.External.Encryption
+import Kernel.Prelude
+import Kernel.Storage.Esqueleto (runTransaction)
+import qualified Kernel.Storage.Esqueleto as Esq
+import qualified Kernel.Types.APISuccess as APISuccess
+import Kernel.Types.Error
+import Kernel.Types.Id
+import Kernel.Types.SlidingWindowLimiter (APIRateLimitOptions)
+import Kernel.Types.Version
+import Kernel.Utils.Common
+import Kernel.Utils.SlidingWindowLimiter
+import Storage.CachedQueries.CacheConfig
+import qualified Storage.CachedQueries.Merchant as QMerchant
+import qualified Storage.Queries.AppInstalls as QAppInstalls
+
+data AppInstallsReq = AppInstallsReq
+  { deviceToken :: Text,
+    source :: Maybe Text,
+    merchantId :: ShortId DMerchant.Merchant,
+    appVersion :: Maybe Version,
+    bundleVersion :: Maybe Version,
+    platform :: Maybe Text
+  }
+  deriving (Generic, ToJSON, FromJSON, ToSchema)
+
+appInstallsHitsCountKey :: Text -> Text
+appInstallsHitsCountKey deviceToken = "BAP:AppInstalls:createAppInstallsDetails:" <> deviceToken <> ":hitsCount"
+
+createAppInstallsDetails ::
+  ( HasFlowEnv m r '["apiRateLimitOptions" ::: APIRateLimitOptions],
+    EsqDBFlow m r,
+    EncFlow m r,
+    CacheFlow m r
+  ) =>
+  AppInstallsReq ->
+  m APISuccess.APISuccess
+createAppInstallsDetails req = do
+  checkSlidingWindowLimit (appInstallsHitsCountKey req.deviceToken)
+  merchant <- QMerchant.findByShortId req.merchantId >>= fromMaybeM (MerchantDoesNotExist req.merchantId.getShortId)
+  mbAppInstalls <- QAppInstalls.findyByAppInstallsAndSource req.deviceToken req.source merchant.id
+  when (isJust mbAppInstalls) $ do
+    Esq.runTransaction $ QAppInstalls.updateAppInstalls req.deviceToken req.source merchant.id
+    throwError $ InvalidRequest "This device token and source already exists."
```

Comment:
Instead of first searching and then updating. We can just have upsert, right? For upsert, you need unique constraint for that I have added comment inside Tabular file.

Reply:
Then you don't need to have all this first search then update if doesn't exist then create. You can just have upsert.